### PR TITLE
[jax 0.7] Support jax 0.7, take 2

### PR DESCRIPTION
**Pull request by PhilipVinc**

### CODE REVIEW

File: netket/jax/_jacobian/jacobian_pytree.py [hunk 1]
```diff
@@ -42,7 +41,9 @@ def jacobian_real_holo(forward_fn: Callable, params: PyTree, samples: Array) ->
     y, vjp_fun = jax.vjp(
         lambda pars: wrap_to_support_scalar(forward_fn)(pars, samples), params
     )
-    (res,) = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
+    # TODO use this once jax bug is fixed
+    # (res,) = vjp_fun(jnp.full_like(y, 1))
+    (res,) = vjp_fun(y * 0 + 1)
```
Comment (PhilipVinc) [inline netket/jax/_jacobian/jacobian_pytree.py:46]: so we loose the 'good type' ?

File: netket/vqs/full_summ/state.py [hunk 2]
```diff
@@ -114,22 +114,6 @@ def __init__(
...
-                )
-            else:
-                par_sharding = jax.sharding.SingleDeviceSharding(jax.devices()[0])
-            variables = jax.tree_util.tree_map(
-                lambda x: jax.lax.with_sharding_constraint(
-                    jnp.asarray(x), par_sharding
-                ),
-                variables,
-            )
-
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/vqs/full_summ/state.py]: Yes, sorry. You are entirely right. 

No, the problem was with sharding.
You might save some checkpoints with `orbax-checkpoint`, which saves sharding information, when running a calculation on 10GPUs and then load it on a single GPU or with only 4.
This loads fine as the mesh is 'abstract' but then fails to execute because you don't have enough GPUs.

Admittedly it is because orbax would require you to provide the new sharding...
I need to find a way to solve it...

But in the meantime let's keep the forced replicated sharding there. We can lift it in a second PR.

File: netket/utils/config_flags.py [hunk 3]
```diff
@@ -295,14 +310,54 @@ def _update_x64(val):
...
+        warnings.warn(
+            f"""
+            NETKET_EXPERIMENTAL_SHARDING mode detected:
+                - SHARDING IS NOW ALWAYS ENABLED, BUT TO USE MORE THAN 1 DEVICE YOU MUST
+                DEFINE THE MESH AND SPECIFY IT IN YOUR CODE.
+
+                - For backward compatibility, specifying `NETKET_EXPERIMENTAL_SHARDING` will
+                create create and set a single-axis mesh with all devices for you.
+
+                - You should UPDATE YOUR CODE to include the following lines, and stop declaring
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/utils/config_flags.py:333]: True. It comes from the previous version of the PR, but indeed this should be changed.

The assumptions of this mode is (1) only a single axis of the mesh and shard the samples and (2) auto mode axes.

[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/utils/config_flags.py:333]: Maybe we just remove the warning for now.

**Comment by PhilipVinc** [comment]
Amazing.
So we can now just wait for jax to release jax 0.7 and we should become happy

File: netket/jax/_jacobian/jacobian_pytree.py [hunk 4]
```diff
@@ -42,7 +41,7 @@ def jacobian_real_holo(forward_fn: Callable, params: PyTree, samples: Array) ->
     y, vjp_fun = jax.vjp(
         lambda pars: wrap_to_support_scalar(forward_fn)(pars, samples), params
     )
-    (res,) = vjp_fun(np.array(1.0, dtype=jnp.result_type(y)))
+    (res,) = vjp_fun(jax.lax.full_like(y, 1.0))
```
Comment (PhilipVinc) [inline netket/jax/_jacobian/jacobian_pytree.py:44]: why? can't we use `jnp.full_like` instead?

Comment (PhilipVinc) [reply netket/jax/_jacobian/jacobian_pytree.py:44]: @inailuig

**Comment by PhilipVinc** [comment]
Assuming tests pass, this could be ready to go.
The only thing I'd like to do is replace the `shard_map` with the new single-axis `jax.smap`, which I think is more correct for what we want to do.
Could you look into it @inailuig ?

File: netket/jax/_jacobian/jacobian_pytree.py (hunk 4 above)
Comment (PhilipVinc) [reply netket/jax/_jacobian/jacobian_pytree.py:44]: It’s ok

File: CHANGELOG.md [hunk 5]
```diff
@@ -38,6 +42,7 @@ The next NetKet version will require Jax 0.7 and Python 3.11 and be incredibly m
 ### Deprecations and Removals
 * {class}`netket.experimental.driver.VMC_SRt` has been deprecated in favour of {class}`netket.experimental.driver.VMC_SR`, which implements both standard SR and the kernel trick/minSR variant, and possibly more efficiently [#2007](https://github.com/netket/netket/pull/2007).
 * Finalized deprecations from August 2024 and earlier [#2108](https://github.com/netket/netket/pull/2108).
+* Constructing an MCState with numpy arrays as variables is deprecated. Please specify jax arrays.
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply CHANGELOG.md:45]: did we not agree to not support it?

File: netket/operator/_kinetic.py [hunk 6]
```diff
@@ -32,9 +32,11 @@ def jacfun(x):
         y, vjp_fun = nkjax.vjp(f, x)
         if y.size == 1:
             eye = jnp.eye(y.size, dtype=x.dtype)[0]
+            eye = jax.lax.pvary(eye, tuple(jax.typeof(y).vma))
             J = jax.vmap(vjp_fun, in_axes=0)(eye)
         else:
             eye = jnp.eye(y.size, dtype=x.dtype)
+            eye = jax.lax.pvary(eye, tuple(jax.typeof(y).vma))
             J = jax.vmap(vjp_fun, in_axes=0)(eye)
         return J
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/operator/_kinetic.py:41]: Is the identity matrix varying though? I would expect it's replicated, and the sharding follows from `x` that is indeed sharded?

[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/operator/_kinetic.py:41]: so what do you think?

[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/operator/_kinetic.py:41]: @inailuig ? should we update this or not? 
I'm not sure we test this in distributed setups...
